### DISCUSSION TITLE
Is there any built-in safe storage api for securely storing secrets?

### DISCUSSION THREAD

**Discussion by HuakunShen** [tauri-apps/tauri#7846] [Ideas]
Electron has a [safe storage](https://www.electronjs.org/docs/latest/api/safe-storage) API for "protecting data stored on disk from being accessed by other applications or users with full disk access".

I need to store a encryption key used to encrypt data on disk. Is there a Tauri or Rust way of doing this on all platforms?

Mac has a [security-framework](https://crates.io/crates/security-framework) for keychain. 

It would be nice if we can have a similar cross-platform plugin for tauri.

**Comment by FabianLars** [comment]
~~Right now, we generally recommend the [stronghold plugin](https://github.com/tauri-apps/plugins-workspace/tree/v1/plugins/stronghold) (or stronghold itself for rust since the plugin doesn't expose a rust api yet) but that does not use the keychain. If you want/need that, there should be rust crates available.~~

stronghold is no longer recommended and will be deprecated and therefore removed in v3.

**Reply by JonasKruckenberg** [reply]
It's also worth noting that stronghold is a way more specialized crate and really only makes sense when you're storing cryptographic key material, for other uses like oauth tokens (which you actually want to be able to retrieve) using the os keychains provides stronger security guarantees 

Edit: And yes there's the [keyring](https://docs.rs/keyring/latest/keyring/) crate which I've used before, we can consider building a secure storage solution using it, but Tauri focus is on getting version 2 done and after that cleaning up tech debt/refactoring to help make contributions easier.

I think in the meantime this would be a great opportunity for a community maintained plugin

**Reply by HuakunShen** [reply]
@JonasKruckenberg An off-topic question: when will V2 be released? I saw a few features in alpha that I really need like the deep linking.

**Reply by FabianLars** [reply]
https://beta.tauri.app/blog/roadmap-to-tauri-2-0/

**Reply by archcorsair** [reply]
Is integrating with keyring still the recommended approach for storing oauth tokens/session information?

**Reply by NoelJacob** [reply]
What about android? keyring doesn't support android

**Reply by Blankeos** [reply]
Does using stronghold imply that your user has to remember a Master password? (I lightly just consulted ChatGPT about this, and it said something about this, is it true?)

**Reply by DomHeadroom** [reply]
> ~Right now, we generally recommend the [stronghold plugin](https://github.com/tauri-apps/plugins-workspace/tree/v1/plugins/stronghold) (or stronghold itself for rust since the plugin doesn't expose a rust api yet) but that does not use the keychain. If you want/need that, there should be rust crates available.~
> 
> stronghold is no longer recommended and will be deprecated and therefore removed in v3.

So, what solution should we use to securly store various sensitive info like auth tokens and so on? (I'm looking for a mobile solution). Is [keyring-rs](https://github.com/open-source-cooperative/keyring-rs) anygood?

**Comment by angelorc** [comment]
Any update on the keyring?

**Reply by sondreb** [reply]
I'm also looking for support for Android.

**Reply by isanjary** [reply]
Any update on the keyring?

**Reply by amitnos123** [reply]
HuakunShen implemted keyring plugin for Tauri V2
https://github.com/HuakunShen/tauri-plugin-keyring

**Reply by GamemakerY** [reply]
Any updates on this? 

It might not be as much of a problem for me, as my backend is in FastAPI, but I would have to change my architecture if I were using Python for storing API keys.

**Comment by belkadimehdi98-commits** [comment]
I ran into a similar challenge while building a Rust + Tauri desktop AI app that stores encrypted session data locally.

One thing I’ve been exploring is a hybrid model:

- Use OS-native secure storage (Keychain / Windows Credential Manager) when available  
- Fall back to encrypted local vault storage (e.g. SQLite + file-level encryption)  
- Keep the encryption key lifecycle separated from app runtime state  

The tricky part seems less about storage itself and more about cross-platform consistency and developer ergonomics.

Curious whether the Tauri core team sees this as something that should live in official plugins long-term, or if the ecosystem model (community plugins like keyring) is the intended direction?

**Reply by VityaSchel** [reply]
you did NOT have to write this comment with ai ✌️
